# Numpy

Install Pandas

pip install pandas

Import Pandas

In [ ]:
import pandas as pd

File Handling

In [ ]:
try:
  data = pd.read_csv('path_of_file')
except FileNotFoundError as e:
  print("File not found:", e)

File not found: [Errno 2] No such file or directory: 'path_of_file'


Read the data

In [ ]:
data = pd.read_csv('/content/drive/MyDrive/Data Science/Python Programming/dataset.csv')

In [ ]:
data

PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
413                            Spector, Mr. Woolf    male   NaN      0      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.0      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket      Fare Cabin Embarked  
0                330911    7.8292   NaN        Q  
1                363272    7.0000   NaN        S  
2                240276    9.6875   NaN        Q  
3                315154    8.6625   NaN        S  
4               3101298   12.2875   NaN        S  
..                  ...       ...   ...      ...  
413           A.5. 3236    8.0500   NaN        S  
414            PC 17758  108.9000  C105        C  
415  SOTON/O.Q. 3101262    7.2500   NaN        S  
416              359309    8.0500   NaN        S  
417                2668   22.3583   NaN        C  

[418 rows x 12 columns]

Display the top 5 records in the data

In [ ]:
data.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

In [ ]:
# DataFrame - it is a kind of datastructure which is only available in panda's
# to access in dataframe type we have to use pandas (pd)
type(data)

pandas.core.frame.DataFrame

How to display fixed value of data like head(15), it will display 15 records

In [ ]:
# display first 15 records
data.head(15)

PassengerId  Survived  Pclass  \
0           892         0       3   
1           893         1       3   
2           894         0       2   
3           895         0       3   
4           896         1       3   
5           897         0       3   
6           898         1       3   
7           899         0       2   
8           900         1       3   
9           901         0       3   
10          902         0       3   
11          903         0       1   
12          904         1       1   
13          905         0       2   
14          906         1       1   

                                                 Name     Sex   Age  SibSp  \
0                                    Kelly, Mr. James    male  34.5      0   
1                    Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2                           Myles, Mr. Thomas Francis    male  62.0      0   
3                                    Wirz, Mr. Albert    male  27.0      0   
4        Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1   
5                          Svensson, Mr. Johan Cervin    male  14.0      0   
6                                Connolly, Miss. Kate  female  30.0      0   
7                        Caldwell, Mr. Albert Francis    male  26.0      1   
8           Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female  18.0      0   
9                             Davies, Mr. John Samuel    male  21.0      2   
10                                   Ilieff, Mr. Ylio    male   NaN      0   
11                         Jones, Mr. Charles Cresson    male  46.0      0   
12      Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0      1   
13                               Howard, Mr. Benjamin    male  63.0      1   
14  Chaffee, Mrs. Herbert Fuller (Carrie Constance...  female  47.0      1   

    Parch       Ticket     Fare Cabin Embarked  
0       0       330911   7.8292   NaN        Q  
1       0       363272   7.0000   NaN        S  
2       0       240276   9.6875   NaN        Q  
3       0       315154   8.6625   NaN        S  
4       1      3101298  12.2875   NaN        S  
5       0         7538   9.2250   NaN        S  
6       0       330972   7.6292   NaN        Q  
7       1       248738  29.0000   NaN        S  
8       0         2657   7.2292   NaN        C  
9       0    A/4 48871  24.1500   NaN        S  
10      0       349220   7.8958   NaN        S  
11      0          694  26.0000   NaN        S  
12      0        21228  82.2667   B45        S  
13      0        24065  26.0000   NaN        S  
14      0  W.E.P. 5734  61.1750   E31        S

In [ ]:
# last 5 records of csv file
data.tail()

PassengerId  Survived  Pclass                          Name     Sex  \
413         1305         0       3            Spector, Mr. Woolf    male   
414         1306         1       1  Oliva y Ocana, Dona. Fermina  female   
415         1307         0       3  Saether, Mr. Simon Sivertsen    male   
416         1308         0       3           Ware, Mr. Frederick    male   
417         1309         0       3      Peter, Master. Michael J    male   

      Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
413   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  39.0      0      0            PC 17758  108.9000  C105        C  
415  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416   NaN      0      0              359309    8.0500   NaN        S  
417   NaN      1      1                2668   22.3583   NaN        C

In [ ]:
# last 15 records of csv file
# Display the last few records in the dataframe
data.tail(15)

PassengerId  Survived  Pclass  \
403         1295         0       1   
404         1296         0       1   
405         1297         0       2   
406         1298         0       2   
407         1299         0       1   
408         1300         1       3   
409         1301         1       3   
410         1302         1       3   
411         1303         1       1   
412         1304         1       3   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                                Name     Sex   Age  SibSp  \
403                           Carrau, Mr. Jose Pedro    male  17.0      0   
404                     Frauenthal, Mr. Isaac Gerald    male  43.0      1   
405     Nourney, Mr. Alfred (Baron von Drachstedt")"    male  20.0      0   
406                        Ware, Mr. William Jeffery    male  23.0      1   
407                       Widener, Mr. George Dunton    male  50.0      1   
408                  Riordan, Miss. Johanna Hannah""  female   NaN      0   
409                        Peacock, Miss. Treasteall  female   3.0      1   
410                           Naughton, Miss. Hannah  female   NaN      0   
411  Minahan, Mrs. William Edward (Lillian E Thorpe)  female  37.0      1   
412                   Henriksson, Miss. Jenny Lovisa  female  28.0      0   
413                               Spector, Mr. Woolf    male   NaN      0   
414                     Oliva y Ocana, Dona. Fermina  female  39.0      0   
415                     Saether, Mr. Simon Sivertsen    male  38.5      0   
416                              Ware, Mr. Frederick    male   NaN      0   
417                         Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
403      0              113059   47.1000   NaN        S  
404      0               17765   27.7208   D40        C  
405      0       SC/PARIS 2166   13.8625   D38        C  
406      0               28666   10.5000   NaN        S  
407      1              113503  211.5000   C80        C  
408      0              334915    7.7208   NaN        Q  
409      1  SOTON/O.Q. 3101315   13.7750   NaN        S  
410      0              365237    7.7500   NaN        Q  
411      0               19928   90.0000   C78        Q  
412      0              347086    7.7750   NaN        S  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C

# Data Structure: Dataframe

In [ ]:
data['PassengerId']

0       892
1       893
2       894
3       895
4       896
       ... 
413    1305
414    1306
415    1307
416    1308
417    1309
Name: PassengerId, Length: 418, dtype: int64

In [ ]:
type(data['PassengerId'])

pandas.core.series.Series

# Series: DataStructure

Series in Depth: It's a one dimensional array-like structure in Pandas that can hold a sequence of value of any data type. Each value in a Series is linked with a associated label, known as the index. You can also customize the index labels in the series.

Create a series from scalar value with the custom index

In [ ]:
series_scalar = pd.Series(10, index=['a', 'b', 'c', 'd', 'e', 'f'])
print(series_scalar)

a    10
b    10
c    10
d    10
e    10
f    10
dtype: int64


In [ ]:
series_scalar1 = pd.Series(10)
print(series_scalar1)

0    10
dtype: int64


In [ ]:
series_scalar['d']

np.int64(10)

Define different value in series as well getting the value with default index numbers.

In [ ]:
data = [10, 20, 30, 40, 50, 60, 70]
series_list = pd.Series(data)
print(series_list)

0    10
1    20
2    30
3    40
4    50
5    60
6    70
dtype: int64


Defining Series values and Index using Numpy Array

In [ ]:
import numpy as np
data_array = np.array([10, 20, 30, 40, 50, 60])
series_custom_array = pd.Series(data_array, index=['a', 'b', 'c', 'd', 'e', 'f'])
series_custom_array

a    10
b    20
c    30
d    40
e    50
f    60
dtype: int64

Defining Series values and Index using Dictionary

In [ ]:
data_dict = {'a': 10, 'b': 20, 'c': 30, 'd': 40, 'e': 50}
series_dict = pd.Series(data_dict)
series_dict

a    10
b    20
c    30
d    40
e    50
dtype: int64

# Slicing

In [ ]:
series_dict[1:4]

b    20
c    30
d    40
dtype: int64

In [ ]:
series_dict['b':'d']

b    20
c    30
d    40
dtype: int64

In [ ]:
series_dict['b':'e']

b    20
c    30
d    40
e    50
dtype: int64

# Attributes in Series Data Structure

In [ ]:
data = ['New Delhi', 'Washington', 'London', 'Paris']
index = ['India', 'USA', 'UK', 'France']
capital_country_series = pd.Series(data, index=index)
print(capital_country_series)

India      New Delhi
USA       Washington
UK            London
France         Paris
dtype: object


In [ ]:
print("Index of the series_capital_country")
print(capital_country_series.index)

Index of the series_capital_country
Index(['India', 'USA', 'UK', 'France'], dtype='object')


In [ ]:
print(type(capital_country_series.index))

<class 'pandas.core.indexes.base.Index'>


In [ ]:
print("Values of the series_capital_country")
print(capital_country_series.values)

Values of the series_capital_country
['New Delhi' 'Washington' 'London' 'Paris']


Dtype is data type of each element in that type of data structure while type discuss about data type of data structure.

In [ ]:
print("Type of the series_capital_country is:", type(capital_country_series))

Type of the series_capital_country is: <class 'pandas.core.series.Series'>


In [ ]:
print("Data Type of the series_capital_country is:", capital_country_series.dtype)

Data Type of the series_capital_country is: object


Shape of Series

In [ ]:
print("Shape of the series_capital_country", capital_country_series.shape)
# 1 demensional array with 4 elements.

Shape of the series_capital_country (4,)


Size of Series

In [ ]:
print("Size of the series_capital_country", capital_country_series.size)

Size of the series_capital_country 4


Empty Series??

In [ ]:
print("Is series_capital_country empty?", capital_country_series.empty)

Is series_capital_country empty? False


Unique elements in Series

In [ ]:
print("Is series_capital_country contains unique elements?", capital_country_series.is_unique)

Is series_capital_country contains unique elements? True


NaN elements in Series?

In [ ]:
print("Is series_capital_country contains NaN elements?", capital_country_series.hasnans)

Is series_capital_country contains NaN elements? False


Dimensions of the Series

In [ ]:
print("Dimensions of the series_capital_country", capital_country_series.ndim)

Dimensions of the series_capital_country 1


How to create a Empty Series?

In [ ]:
empty_series = pd.Series()
print(empty_series)

Series([], dtype: object)


In [ ]:
print("Is empty_series empty?")
print(empty_series.empty)

Is empty_series empty?
True


# Methods in Series Data Structure

In [ ]:
capital_country_series.head(2)

India     New Delhi
USA      Washington
dtype: object

In [ ]:
capital_country_series.tail(2)

UK        London
France     Paris
dtype: object

In [ ]:
marks = [85, 60, 74, 66, 89, 91]
students = ['Shahnawaz', 'Ajmal', 'Junaid', 'Waleed', 'Akbar', 'Shahid']

series_students = pd.Series(marks, index=students)
print(series_students)

Shahnawaz    85
Ajmal        60
Junaid       74
Waleed       66
Akbar        89
Shahid       91
dtype: int64


To get the descriptive statistics of the given data like Mean, Minimum, Maximum

In [ ]:
series_students.describe()

count     6.000000
mean     77.500000
std      12.817956
min      60.000000
25%      68.000000
50%      79.500000
75%      88.000000
max      91.000000
dtype: float64

value_counts() - Unique value in data, and frequency or count of that data. Helps in Classification problems - We have labelled data of two classes and we have to check unique values in both classes wrt features/columns.

value_counts()

In [ ]:
series_students.value_counts()

85    1
60    1
74    1
66    1
89    1
91    1
Name: count, dtype: int64

sort_values()

In [ ]:
series_students.sort_values()

Ajmal        60
Waleed       66
Junaid       74
Shahnawaz    85
Akbar        89
Shahid       91
dtype: int64

sort_index()

In [ ]:
series_students.sort_index()

Ajmal        60
Akbar        89
Junaid       74
Shahid       91
Shahnawaz    85
Waleed       66
dtype: int64

drop()

In [ ]:
# Real Changes in Internal Data Structure Data
series_students = series_students.drop('Waleed')
series_students

Shahnawaz    85
Junaid       74
Akbar        89
Shahid       91
dtype: int64

In [ ]:
# Real Changes in Internal Data Structure Data
series_students.drop('Akbar', inplace=True)
series_students

Shahnawaz    85
Junaid       74
Shahid       91
dtype: int64

In [ ]:
# no changes in internal data structure data
series_students.drop('Shahnawaz')

Junaid    74
Shahid    91
dtype: int64

In [ ]:
series_students

Shahnawaz    85
Junaid       74
Shahid       91
dtype: int64

isnull() -> Detect the missing values

In [ ]:
print(series_students.isnull())

Shahnawaz    False
Junaid       False
Shahid       False
dtype: bool


to sum the all values in series data structure

In [ ]:
print(series_students.isnull().sum())

0


To Add any value back in the data.

In [ ]:
series_students['Ajmal'] = 91
series_students

Shahnawaz    85
Junaid       74
Shahid       91
Ajmal        91
dtype: int64

Task: Mathematical Operations on Series


1.   Addition of two Series
2.   Subtraction of two Series
3.   Multiplication of two Series
4.   Division of two Series

In [ ]:
import pandas as pd

# Create two Series
s1 = pd.Series([10, 20, 30, 40])
s2 = pd.Series([1, 2, 3, 4])

# Add them
result = s1 + s2
print(result)

# multiply them
multiply_result = s1 * s2
print(multiply_result)

# subtract them
subtract_result = s1 - s2
print(subtract_result)

# divide them
divide_result = s1 / s2
print(divide_result)

0    11
1    22
2    33
3    44
dtype: int64
0     10
1     40
2     90
3    160
dtype: int64
0     9
1    18
2    27
3    36
dtype: int64
0    10.0
1    10.0
2    10.0
3    10.0
dtype: float64


Task: Simulate the same approach for a dataframe and you need to cover all the attributes and methods for the dataframe

Import all the required libraries

# ADVANCED PANDA'S

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
data = pd.read_excel('/content/drive/MyDrive/Data Science/Python Programming/customers.xlsx')
data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  
0   Trailsway          Road    San Diego  California   92127.0  
1         NaN           NaN      El Paso       Texas       NaN  
2    Arkansas        Circle  San Antonio       Texas   78220.0  
3        Debs         Point      Memphis   Tennessee   38150.0  
4         2nd         Drive          NaN         NaN   83732.0

Information about Customers Dataframe

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             1000 non-null   int64  
 1   first_name     1000 non-null   object 
 2   last_name      1000 non-null   object 
 3   email          878 non-null    object 
 4   gender         957 non-null    object 
 5   street_num     738 non-null    float64
 6   street_name    963 non-null    object 
 7   street_suffix  963 non-null    object 
 8   city           921 non-null    object 
 9   state          920 non-null    object 
 10  postcode       843 non-null    float64
dtypes: float64(2), int64(1), object(8)
memory usage: 86.1+ KB


In [ ]:
data.isnull()

id  first_name  last_name  email  gender  street_num  street_name  \
0    False       False      False  False   False       False        False   
1    False       False      False  False   False        True         True   
2    False       False      False  False   False       False        False   
3    False       False      False  False   False       False        False   
4    False       False      False  False    True        True        False   
..     ...         ...        ...    ...     ...         ...          ...   
995  False       False      False  False   False       False        False   
996  False       False      False  False   False       False        False   
997  False       False      False   True   False       False        False   
998  False       False      False  False   False        True        False   
999  False       False      False   True   False       False        False   

     street_suffix   city  state  postcode  
0            False  False  False     False  
1             True  False  False      True  
2            False  False  False     False  
3            False  False  False     False  
4            False   True   True     False  
..             ...    ...    ...       ...  
995          False  False  False     False  
996          False  False  False     False  
997          False  False  False     False  
998          False  False  False     False  
999          False  False  False     False  

[1000 rows x 11 columns]

Add a new column named as 'Country' and the value inside that is 'USA' in the above dataframe.

In [ ]:
data['Country'] = 'USA'

In [ ]:
data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode Country  
0   Trailsway          Road    San Diego  California   92127.0     USA  
1         NaN           NaN      El Paso       Texas       NaN     USA  
2    Arkansas        Circle  San Antonio       Texas   78220.0     USA  
3        Debs         Point      Memphis   Tennessee   38150.0     USA  
4         2nd         Drive          NaN         NaN   83732.0     USA

Task 1: I want to replace/::place the 'country' column after the 'city' column.

In [ ]:
data = data.reindex(columns=[
    *data.columns[:data.columns.get_loc('city')],
    'Country',
    'city',
    *data.columns[data.columns.get_loc('city') + 1:]
])

data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix Country         city       state  postcode Country  
0   Trailsway          Road     USA    San Diego  California   92127.0     USA  
1         NaN           NaN     USA      El Paso       Texas       NaN     USA  
2    Arkansas        Circle     USA  San Antonio       Texas   78220.0     USA  
3        Debs         Point     USA      Memphis   Tennessee   38150.0     USA  
4         2nd         Drive     USA          NaN         NaN   83732.0     USA

append: Add column in the end of the table.

In [ ]:
# we can only add one value of the column or complete total records values like here 1000 index
membership_status = ['Gold', 'Silver', 'Bronze']
data['membership_status'] = membership_status
data.head()

ValueError: Length of values (3) does not match length of index (1000)

Create a new column 'full_name' which is composed of first_name and the last_name.

In [ ]:
data['full_name'] = data['first_name'] + ' ' + data['last_name']
data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode Country  \
0   Trailsway          Road    San Diego  California   92127.0     USA   
1         NaN           NaN      El Paso       Texas       NaN     USA   
2    Arkansas        Circle  San Antonio       Texas   78220.0     USA   
3        Debs         Point      Memphis   Tennessee   38150.0     USA   
4         2nd         Drive          NaN         NaN   83732.0     USA   

           full_name  
0   Romain Southcott  
1   Cosimo Molyneaux  
2      Bambi Westrip  
3  Roarke Pankettman  
4   Mikaela Althorpe

In [ ]:
# this is a wrong way to compose/concate as new column is not created where we
# will add both columns
concate_first_last = data['first_name' + 'last_name']
data.head()

KeyError: 'first_namelast_name'

Access the rows in a given dataframe: .loc() vs .iloc()

In [ ]:
dataframe_demo = {
    'Name': ['Shahnawaz', 'Amir', 'Waleed', 'Akbar', 'Junaid'],
    'Age': [23, 34, 25, 27, 36],
    'City': ['Rawalpindi', 'Islamabad', 'Hyderabad', 'Lahore', 'Karachi']
}

# initializing index to values
dataframe_demo = pd.DataFrame(dataframe_demo, index=['a', 'b', 'c', 'd', 'e'])
dataframe_demo

Name  Age        City
a  Shahnawaz   23  Rawalpindi
b       Amir   34   Islamabad
c     Waleed   25   Hyderabad
d      Akbar   27      Lahore
e     Junaid   36     Karachi

In [ ]:
# work on customised indexes
# last index is inclusive
dataframe_demo.loc['b']

Name         Amir
Age            34
City    Islamabad
Name: b, dtype: object

In [ ]:
dataframe_demo.loc['b':'d']

Name  Age       City
b    Amir   34  Islamabad
c  Waleed   25  Hyderabad
d   Akbar   27     Lahore

In [ ]:
dataframe_demo.loc[0]

KeyError: 0

In [ ]:
# ending index is excluded/exclusive
dataframe_demo.iloc[1]

Name         Amir
Age            34
City    Islamabad
Name: b, dtype: object

In [ ]:
dataframe_demo.iloc[1:4]

Name  Age       City
b    Amir   34  Islamabad
c  Waleed   25  Hyderabad
d   Akbar   27     Lahore

In [ ]:
dataframe_demo_v1 = {
    'Name': ['Shahnawaz', 'Amir', 'Waleed', 'Akbar', 'Junaid'],
    'Age': [23, 34, 25, 27, 36],
    'City': ['Rawalpindi', 'Islamabad', 'Hyderabad', 'Lahore', 'Karachi']
}

dataframe_demo_v1 = pd.DataFrame(dataframe_demo_v1)
dataframe_demo_v1

Name  Age        City
0  Shahnawaz   23  Rawalpindi
1       Amir   34   Islamabad
2     Waleed   25   Hyderabad
3      Akbar   27      Lahore
4     Junaid   36     Karachi

In [ ]:
dataframe_demo_v1.loc[0]

Name     Shahnawaz
Age             23
City    Rawalpindi
Name: 0, dtype: object

In [ ]:
dataframe_demo_v1.iloc[0]

Name     Shahnawaz
Age             23
City    Rawalpindi
Name: 0, dtype: object

In [ ]:
data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode Country  \
0   Trailsway          Road    San Diego  California   92127.0     USA   
1         NaN           NaN      El Paso       Texas       NaN     USA   
2    Arkansas        Circle  San Antonio       Texas   78220.0     USA   
3        Debs         Point      Memphis   Tennessee   38150.0     USA   
4         2nd         Drive          NaN         NaN   83732.0     USA   

           full_name  
0   Romain Southcott  
1   Cosimo Molyneaux  
2      Bambi Westrip  
3  Roarke Pankettman  
4   Mikaela Althorpe

In [ ]:
data.shape

(1000, 13)

Access the details of the id's of the customers from 500 to 504.

In [ ]:
data.loc[500:504]

id first_name last_name                       email gender  street_num  \
500  501     Elliot  Polycote  epolycotedw@infoseek.co.jp   Male         NaN   
501  502   Shirline    Halpin     shalpindx@moonfruit.com    NaN     76640.0   
502  503      Darcy  Cruddace                         NaN    NaN         NaN   
503  504    Toiboid   Bentzen     tbentzendz@freewebs.com   Male       913.0   
504  505      Paxon    Corris       pcorrise0@behance.net   Male      8362.0   

    street_name street_suffix          city        state  postcode Country  \
500    Prentice         Drive      Hartford  Connecticut       NaN     USA   
501       Buell        Circle        Dallas        Texas   75216.0     USA   
502   Ridgeview         Court         Utica     New York   13505.0     USA   
503         2nd         Trail  White Plains     New York   10606.0     USA   
504       Eliot           Way    Des Moines         Iowa   50362.0     USA   

           full_name  
500  Elliot Polycote  
501  Shirline Halpin  
502   Darcy Cruddace  
503  Toiboid Bentzen  
504     Paxon Corris

In [ ]:
data.iloc[500:504]

id first_name last_name                       email gender  street_num  \
500  501     Elliot  Polycote  epolycotedw@infoseek.co.jp   Male         NaN   
501  502   Shirline    Halpin     shalpindx@moonfruit.com    NaN     76640.0   
502  503      Darcy  Cruddace                         NaN    NaN         NaN   
503  504    Toiboid   Bentzen     tbentzendz@freewebs.com   Male       913.0   

    street_name street_suffix          city        state  postcode Country  \
500    Prentice         Drive      Hartford  Connecticut       NaN     USA   
501       Buell        Circle        Dallas        Texas   75216.0     USA   
502   Ridgeview         Court         Utica     New York   13505.0     USA   
503         2nd         Trail  White Plains     New York   10606.0     USA   

           full_name  
500  Elliot Polycote  
501  Shirline Halpin  
502   Darcy Cruddace  
503  Toiboid Bentzen

In [ ]:
data[(data['id'] >= 500) & (data['id'] <= 504)]

id first_name last_name                       email  gender  street_num  \
499  500   Carolann   Balcock        cbalcockdv@alexa.com  Female         NaN   
500  501     Elliot  Polycote  epolycotedw@infoseek.co.jp    Male         NaN   
501  502   Shirline    Halpin     shalpindx@moonfruit.com     NaN     76640.0   
502  503      Darcy  Cruddace                         NaN     NaN         NaN   
503  504    Toiboid   Bentzen     tbentzendz@freewebs.com    Male       913.0   

      street_name street_suffix          city        state  postcode Country  \
499  Stone Corner          Pass       El Paso        Texas   79916.0     USA   
500      Prentice         Drive      Hartford  Connecticut       NaN     USA   
501         Buell        Circle        Dallas        Texas   75216.0     USA   
502     Ridgeview         Court         Utica     New York   13505.0     USA   
503           2nd         Trail  White Plains     New York   10606.0     USA   

            full_name  
499  Carolann Balcock  
500   Elliot Polycote  
501   Shirline Halpin  
502    Darcy Cruddace  
503   Toiboid Bentzen

In [ ]:
# iloc = when we know row/column position = last index is exclusive
# loc = when we have labelled customised index = last index is inclusive
data[500:504]

id first_name last_name                       email gender  street_num  \
500  501     Elliot  Polycote  epolycotedw@infoseek.co.jp   Male         NaN   
501  502   Shirline    Halpin     shalpindx@moonfruit.com    NaN     76640.0   
502  503      Darcy  Cruddace                         NaN    NaN         NaN   
503  504    Toiboid   Bentzen     tbentzendz@freewebs.com   Male       913.0   

    street_name street_suffix          city        state  postcode Country  \
500    Prentice         Drive      Hartford  Connecticut       NaN     USA   
501       Buell        Circle        Dallas        Texas   75216.0     USA   
502   Ridgeview         Court         Utica     New York   13505.0     USA   
503         2nd         Trail  White Plains     New York   10606.0     USA   

           full_name  
500  Elliot Polycote  
501  Shirline Halpin  
502   Darcy Cruddace  
503  Toiboid Bentzen

In [ ]:
del data['Country']

In [ ]:
data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  \
0   Trailsway          Road    San Diego  California   92127.0   
1         NaN           NaN      El Paso       Texas       NaN   
2    Arkansas        Circle  San Antonio       Texas   78220.0   
3        Debs         Point      Memphis   Tennessee   38150.0   
4         2nd         Drive          NaN         NaN   83732.0   

           full_name  
0   Romain Southcott  
1   Cosimo Molyneaux  
2      Bambi Westrip  
3  Roarke Pankettman  
4   Mikaela Althorpe

Add country column again

In [ ]:
data['Country'] = 'USA'

In [ ]:
data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  \
0   Trailsway          Road    San Diego  California   92127.0   
1         NaN           NaN      El Paso       Texas       NaN   
2    Arkansas        Circle  San Antonio       Texas   78220.0   
3        Debs         Point      Memphis   Tennessee   38150.0   
4         2nd         Drive          NaN         NaN   83732.0   

           full_name Country  
0   Romain Southcott     USA  
1   Cosimo Molyneaux     USA  
2      Bambi Westrip     USA  
3  Roarke Pankettman     USA  
4   Mikaela Althorpe     USA

In [ ]:
# Same end result as del data['Country'], but done using a panda's method
data.drop(columns=['Country'], inplace=True)
data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  \
0   Trailsway          Road    San Diego  California   92127.0   
1         NaN           NaN      El Paso       Texas       NaN   
2    Arkansas        Circle  San Antonio       Texas   78220.0   
3        Debs         Point      Memphis   Tennessee   38150.0   
4         2nd         Drive          NaN         NaN   83732.0   

           full_name  
0   Romain Southcott  
1   Cosimo Molyneaux  
2      Bambi Westrip  
3  Roarke Pankettman  
4   Mikaela Althorpe

Rename the row index with the custom index.

'0' - 'firstCustomer'

'1' - 'secondCustomer'

In [ ]:
rename_dict = {
    0: 'firstCustomer',
    1: 'secondCustomer'
}

data.rename(index=rename_dict, inplace=True)
data.head()

id first_name   last_name                      email  gender  \
firstCustomer    1     Romain   Southcott  rsouthcott0@clickbank.net    Male   
secondCustomer   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male   
2                3      Bambi     Westrip     bwestrip2@symantec.com  Female   
3                4     Roarke  Pankettman     rpankettman3@wiley.com    Male   
4                5    Mikaela    Althorpe           malthorpe4@51.la    Male   

                street_num street_name street_suffix         city       state  \
firstCustomer          1.0   Trailsway          Road    San Diego  California   
secondCustomer         NaN         NaN           NaN      El Paso       Texas   
2                   4057.0    Arkansas        Circle  San Antonio       Texas   
3                     74.0        Debs         Point      Memphis   Tennessee   
4                      NaN         2nd         Drive          NaN         NaN   

                postcode  
firstCustomer    92127.0  
secondCustomer       NaN  
2                78220.0  
3                38150.0  
4                83732.0

In [ ]:
# Because .loc looks for labels 0 and 1, not positions.
# .loc take customised positions while .iloc take default positions.
# For non customised index iloc = loc else they are different.
data.loc[0:1]

TypeError: cannot do slice indexing on Index with these indexers [0] of type int

In [ ]:
data.iloc[0:1]

id first_name  last_name                      email gender  \
firstCustomer   1     Romain  Southcott  rsouthcott0@clickbank.net   Male   

               street_num street_name street_suffix       city       state  \
firstCustomer         1.0   Trailsway          Road  San Diego  California   

               postcode         full_name  
firstCustomer   92127.0  Romain Southcott

In [ ]:
data.loc['firstCustomer':'secondCustomer']

id first_name  last_name                      email gender  \
firstCustomer    1     Romain  Southcott  rsouthcott0@clickbank.net   Male   
secondCustomer   2     Cosimo  Molyneaux      cmolyneaux1@wiley.com   Male   

                street_num street_name street_suffix       city       state  \
firstCustomer          1.0   Trailsway          Road  San Diego  California   
secondCustomer         NaN         NaN           NaN    El Paso       Texas   

                postcode         full_name  
firstCustomer    92127.0  Romain Southcott  
secondCustomer       NaN  Cosimo Molyneaux

Finding the unique values

In [ ]:
unique_genders = data['gender'].unique()
print(unique_genders)

['Male' 'Female' nan]


Count of the unique values in the given column in a dataframe.

Note: It will only return the non-null values as the count.

In [ ]:
count_unique_gender = data['gender'].nunique()
print(count_unique_gender)

2


In [ ]:
data['gender'].value_counts()

gender
Male      485
Female    472
Name: count, dtype: int64

In [ ]:
data.dtypes

id                 int64
first_name        object
last_name         object
email             object
gender            object
street_num       float64
street_name       object
street_suffix     object
Country           object
city              object
state             object
postcode         float64
Country           object
dtype: object

Task 2: Can you rename the columns of street_num with the street_number and street_name with the street_address.

In [ ]:
rename_column = {
    'street_num': 'street_number',
    'street_name': 'street_address'
}

data.rename(columns=rename_column, inplace=True)
data.head()

id first_name   last_name                      email  gender  \
firstCustomer    1     Romain   Southcott  rsouthcott0@clickbank.net    Male   
secondCustomer   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male   
2                3      Bambi     Westrip     bwestrip2@symantec.com  Female   
3                4     Roarke  Pankettman     rpankettman3@wiley.com    Male   
4                5    Mikaela    Althorpe           malthorpe4@51.la     NaN   

                street_number street_address street_suffix         city  \
firstCustomer             1.0      Trailsway          Road    San Diego   
secondCustomer            NaN            NaN           NaN      El Paso   
2                      4057.0       Arkansas        Circle  San Antonio   
3                        74.0           Debs         Point      Memphis   
4                         NaN            2nd         Drive          NaN   

                     state  postcode          full_name  
firstCustomer   California   92127.0   Romain Southcott  
secondCustomer       Texas       NaN   Cosimo Molyneaux  
2                    Texas   78220.0      Bambi Westrip  
3                Tennessee   38150.0  Roarke Pankettman  
4                      NaN   83732.0   Mikaela Althorpe

# Purchases

In [ ]:
data_purchases = pd.read_excel('/content/drive/MyDrive/Data Science/Python Programming/purchases.xlsx')
data_purchases.head()

Unnamed: 0  id           purch_date  customer_num  product_num  amount  \
0           0   1  2019-01-03 00:00:00           823           27      12   
1           1   2  2019-01-03 00:00:00           606           28      14   
2           2   3  2019-01-03 00:00:00           955            9      17   
3           3   4  2019-01-03 00:00:00           577           19       3   
4           4   5  2019-01-03 00:00:00           429            8      18   

      paid  
0  $568.92  
1  $395.36  
2  $510.17  
3   $68.49  
4  $759.42

In [ ]:
data_purchases.dtypes

Unnamed: 0       int64
id               int64
purch_date      object
customer_num     int64
product_num      int64
amount           int64
paid            object
dtype: object

Delete a column from the data set

In [ ]:
data_purchases.drop(columns=['Unnamed: 0'], inplace=True)
data_purchases.head()

id           purch_date  customer_num  product_num  amount     paid
0   1  2019-01-03 00:00:00           823           27      12  $568.92
1   2  2019-01-03 00:00:00           606           28      14  $395.36
2   3  2019-01-03 00:00:00           955            9      17  $510.17
3   4  2019-01-03 00:00:00           577           19       3   $68.49
4   5  2019-01-03 00:00:00           429            8      18  $759.42

Handle the data types of the column in an appropriate manner.

purch_data -> datetime

In [ ]:
data_purchases['purch_date'] = pd.to_datetime(data_purchases['purch_date'])
data_purchases.dtypes

id                       int64
purch_date      datetime64[ns]
customer_num             int64
product_num              int64
amount                   int64
paid                    object
dtype: object

id -> object

In [ ]:
data_purchases['id'] = data_purchases['id'].astype(object)
data_purchases.dtypes

id                      object
purch_date      datetime64[ns]
customer_num             int64
product_num              int64
amount                   int64
paid                    object
dtype: object

paid -> float

In [ ]:
## paid -> data preprocess and then only we can do the type-conversion

## the below code will simply remove the $ and , from the paid column using a regular expression
data_purchases['paid'] = data_purchases['paid'].str.replace('[\$,]', '', regex=True)
data_purchases.head()

<>:4: SyntaxWarning: invalid escape sequence '\$'
<>:4: SyntaxWarning: invalid escape sequence '\$'
/tmp/ipython-input-148438409.py:4: SyntaxWarning: invalid escape sequence '\$'
  data_purchases['paid'] = data_purchases['paid'].str.replace('[\$,]', '', regex=True)


id purch_date  customer_num  product_num  amount    paid
0  1 2019-01-03           823           27      12  568.92
1  2 2019-01-03           606           28      14  395.36
2  3 2019-01-03           955            9      17  510.17
3  4 2019-01-03           577           19       3   68.49
4  5 2019-01-03           429            8      18  759.42

In [ ]:
# in panda's object is basically a string
data_purchases['paid'] = data_purchases['paid'].astype(float)
data_purchases.dtypes

id                      object
purch_date      datetime64[ns]
customer_num             int64
product_num              int64
amount                   int64
paid                   float64
dtype: object

# Products

In [ ]:
data_products = pd.read_excel('/content/drive/MyDrive/Data Science/Python Programming/products.xlsx')
data_products.head()

Unnamed: 0  id                           product    cost   company
0           0   1              Liners - Baking Cups   $6.36  Skipfire
1           1   2        Nori Sea Weed - Gold Label  $85.74   Dynazzy
2           2   3                Bar Bran Honey Nut  $65.40      Ntag
3           3   4        Soup - Campbells Beef Stew  $68.16  Photojam
4           4   5  Wine - Shiraz Wolf Blass Premium  $87.39      Eare

In [ ]:
data_products.drop(columns=['Unnamed: 0'], inplace=True)
data_products.head()

id                           product    cost   company
0   1              Liners - Baking Cups   $6.36  Skipfire
1   2        Nori Sea Weed - Gold Label  $85.74   Dynazzy
2   3                Bar Bran Honey Nut  $65.40      Ntag
3   4        Soup - Campbells Beef Stew  $68.16  Photojam
4   5  Wine - Shiraz Wolf Blass Premium  $87.39      Eare

In [ ]:
data_products['id'] = data_products['id'].astype(object)
data_products.dtypes

id         object
product    object
cost       object
company    object
dtype: object

In [ ]:
## the below code will simply remove the $ and , from the paid column using a regular expression
data_products['cost'] = data_products['cost'].str.replace('[\$,]', '', regex=True)
data_products.head()

<>:2: SyntaxWarning: invalid escape sequence '\$'
<>:2: SyntaxWarning: invalid escape sequence '\$'
/tmp/ipython-input-797701294.py:2: SyntaxWarning: invalid escape sequence '\$'
  data_products['cost'] = data_products['cost'].str.replace('[\$,]', '', regex=True)


id                           product   cost   company
0  1              Liners - Baking Cups   6.36  Skipfire
1  2        Nori Sea Weed - Gold Label  85.74   Dynazzy
2  3                Bar Bran Honey Nut  65.40      Ntag
3  4        Soup - Campbells Beef Stew  68.16  Photojam
4  5  Wine - Shiraz Wolf Blass Premium  87.39      Eare

In [ ]:
# in panda's object is basically a string
data_products['cost'] = data_products['cost'].astype(float)
data_products.dtypes

id          object
product     object
cost       float64
company     object
dtype: object

In [ ]:
data_purchases.shape

(6000, 7)

In [ ]:
data_products.shape

(60, 5)

In [ ]:
data.shape

(1000, 11)

# SQL

SQL Union

pd.concat -> to manipulation and combine multiple dataframes in pandas.

In [ ]:
## axis = 0[default] row-wise concatenation,
## axis = 1 -> column wise concatenation
combined_df = pd.concat([data, data_purchases, data_products])
combined_df

id first_name   last_name                      email  gender  street_num  \
0    1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1    2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2    3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3    4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4    5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   
..  ..        ...         ...                        ...     ...         ...   
55  56        NaN         NaN                        NaN     NaN         NaN   
56  57        NaN         NaN                        NaN     NaN         NaN   
57  58        NaN         NaN                        NaN     NaN         NaN   
58  59        NaN         NaN                        NaN     NaN         NaN   
59  60        NaN         NaN                        NaN     NaN         NaN   

   street_name street_suffix         city       state  postcode  Unnamed: 0  \
0    Trailsway          Road    San Diego  California   92127.0         NaN   
1          NaN           NaN      El Paso       Texas       NaN         NaN   
2     Arkansas        Circle  San Antonio       Texas   78220.0         NaN   
3         Debs         Point      Memphis   Tennessee   38150.0         NaN   
4          2nd         Drive          NaN         NaN   83732.0         NaN   
..         ...           ...          ...         ...       ...         ...   
55         NaN           NaN          NaN         NaN       NaN        55.0   
56         NaN           NaN          NaN         NaN       NaN        56.0   
57         NaN           NaN          NaN         NaN       NaN        57.0   
58         NaN           NaN          NaN         NaN       NaN        58.0   
59         NaN           NaN          NaN         NaN       NaN        59.0   

   purch_date  customer_num  product_num  amount paid  \
0         NaN           NaN          NaN     NaN  NaN   
1         NaN           NaN          NaN     NaN  NaN   
2         NaN           NaN          NaN     NaN  NaN   
3         NaN           NaN          NaN     NaN  NaN   
4         NaN           NaN          NaN     NaN  NaN   
..        ...           ...          ...     ...  ...   
55        NaN           NaN          NaN     NaN  NaN   
56        NaN           NaN          NaN     NaN  NaN   
57        NaN           NaN          NaN     NaN  NaN   
58        NaN           NaN          NaN     NaN  NaN   
59        NaN           NaN          NaN     NaN  NaN   

                          product    cost   company  
0                             NaN     NaN       NaN  
1                             NaN     NaN       NaN  
2                             NaN     NaN       NaN  
3                             NaN     NaN       NaN  
4                             NaN     NaN       NaN  
..                            ...     ...       ...  
55                  Peach - Fresh  $20.53     Zazio  
56            Muffin Hinge - 211n  $68.63  Fanoodle  
57             Bar Bran Honey Nut  $73.05    Yakijo  
58  Nut - Almond, Blanched, Whole  $74.28     Eazzy  
59       Table Cloth 90x90 Colour  $41.22     Quinu  

[7060 rows x 20 columns]

SQL Joins => Inner, Left, Right, Outer

pd.merge

In [ ]:
data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  
0   Trailsway          Road    San Diego  California   92127.0  
1         NaN           NaN      El Paso       Texas       NaN  
2    Arkansas        Circle  San Antonio       Texas   78220.0  
3        Debs         Point      Memphis   Tennessee   38150.0  
4         2nd         Drive          NaN         NaN   83732.0

In [ ]:
data_purchases.head()

Unnamed: 0  id           purch_date  customer_num  product_num  amount  \
0           0   1  2019-01-03 00:00:00           823           27      12   
1           1   2  2019-01-03 00:00:00           606           28      14   
2           2   3  2019-01-03 00:00:00           955            9      17   
3           3   4  2019-01-03 00:00:00           577           19       3   
4           4   5  2019-01-03 00:00:00           429            8      18   

      paid  
0  $568.92  
1  $395.36  
2  $510.17  
3   $68.49  
4  $759.42

Inner Join

In [ ]:
merge_data = pd.merge(data, data_purchases, on='id', how='inner')
merge_data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  Unnamed: 0  \
0   Trailsway          Road    San Diego  California   92127.0           0   
1         NaN           NaN      El Paso       Texas       NaN           1   
2    Arkansas        Circle  San Antonio       Texas   78220.0           2   
3        Debs         Point      Memphis   Tennessee   38150.0           3   
4         2nd         Drive          NaN         NaN   83732.0           4   

            purch_date  customer_num  product_num  amount     paid  
0  2019-01-03 00:00:00           823           27      12  $568.92  
1  2019-01-03 00:00:00           606           28      14  $395.36  
2  2019-01-03 00:00:00           955            9      17  $510.17  
3  2019-01-03 00:00:00           577           19       3   $68.49  
4  2019-01-03 00:00:00           429            8      18  $759.42

In [ ]:
merge_data.shape

(1000, 17)

Right Join

In [ ]:
merge_data = pd.merge(data, data_purchases, on='id', how='right')
merge_data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  Unnamed: 0  \
0   Trailsway          Road    San Diego  California   92127.0           0   
1         NaN           NaN      El Paso       Texas       NaN           1   
2    Arkansas        Circle  San Antonio       Texas   78220.0           2   
3        Debs         Point      Memphis   Tennessee   38150.0           3   
4         2nd         Drive          NaN         NaN   83732.0           4   

            purch_date  customer_num  product_num  amount     paid  
0  2019-01-03 00:00:00           823           27      12  $568.92  
1  2019-01-03 00:00:00           606           28      14  $395.36  
2  2019-01-03 00:00:00           955            9      17  $510.17  
3  2019-01-03 00:00:00           577           19       3   $68.49  
4  2019-01-03 00:00:00           429            8      18  $759.42

In [ ]:
merge_data.shape

(6000, 17)

Left Join

In [ ]:
merge_data = pd.merge(data, data_purchases, on='id', how='left')
merge_data.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  Unnamed: 0  \
0   Trailsway          Road    San Diego  California   92127.0           0   
1         NaN           NaN      El Paso       Texas       NaN           1   
2    Arkansas        Circle  San Antonio       Texas   78220.0           2   
3        Debs         Point      Memphis   Tennessee   38150.0           3   
4         2nd         Drive          NaN         NaN   83732.0           4   

            purch_date  customer_num  product_num  amount     paid  
0  2019-01-03 00:00:00           823           27      12  $568.92  
1  2019-01-03 00:00:00           606           28      14  $395.36  
2  2019-01-03 00:00:00           955            9      17  $510.17  
3  2019-01-03 00:00:00           577           19       3   $68.49  
4  2019-01-03 00:00:00           429            8      18  $759.42

In [ ]:
merge_data.shape

(1000, 17)

In [ ]:
data1_customers = pd.DataFrame({
    'id': [1, 2, 3],
    'name': ['alice', 'bob', 'charlie'],
    'product':['book', 'pen', 'notebook']
})


data2_purchases = pd.DataFrame({
    'id': [2, 3, 4],
    'product': ['pen', 'notebook', 'pen'],
    'amount' : [5, 10, 15]
})

# now we want entire information of the customers details, including amount as well
# if we want to get common id's we will use inner join
merge_data = pd.merge(data1_customers, data2_purchases, on='id', how='inner')
merge_data.head()

id     name product_x product_y  amount
0   2      bob       pen       pen       5
1   3  charlie  notebook  notebook      10

In [ ]:
# if we want to get common id's and data1_customers records as well, we will use left join
merge_data = pd.merge(data1_customers, data2_purchases, on='id', how='left')
merge_data.head()

id     name product_x product_y  amount
0   1    alice      book       NaN     NaN
1   2      bob       pen       pen     5.0
2   3  charlie  notebook  notebook    10.0

In [ ]:
# if we want to get common id's and data2_purchases records as well, we will use right join
merge_data = pd.merge(data1_customers, data2_purchases, on='id', how='right')
merge_data.head()

id     name product_x product_y  amount
0   2      bob       pen       pen       5
1   3  charlie  notebook  notebook      10
2   4      NaN       NaN       pen      15

 suffixes=('_cust', '_pur'))

In [ ]:
merge_data = pd.merge(data1_customers, data2_purchases, on='id', how='inner', suffixes=('_cust', '_pur'))
merge_data.head()

id     name product_cust product_pur  amount
0   2      bob          pen         pen       5
1   3  charlie     notebook    notebook      10

Combine both Data in One

In [ ]:
combined_df = pd.concat([data1_customers, data2_purchases])
combined_df

id     name   product  amount
0   1    alice      book     NaN
1   2      bob       pen     NaN
2   3  charlie  notebook     NaN
0   2      NaN       pen     5.0
1   3      NaN  notebook    10.0
2   4      NaN       pen    15.0

Task 3: Difference between pd.concat, pd.merge and pd.join

pd.concat() – Combine DataFrames Along Rows or Columns

In [ ]:
import pandas as pd

df1 = pd.DataFrame({
    'ID': [1, 2, 3],
    'Name': ['Ali', 'Sara', 'Omar']
})

df2 = pd.DataFrame({
    'ID': [4, 5],
    'Name': ['Zara', 'Hassan']
})

# Combine vertically (default: axis=0)
result = pd.concat([df1, df2])
print(result)

   ID    Name
0   1     Ali
1   2    Sara
2   3    Omar
0   4    Zara
1   5  Hassan


pd.merge() – Combine DataFrames Based on a Common Column (SQL-style JOIN)

In [ ]:
employees = pd.DataFrame({
    'EmpID': [1, 2, 3],
    'Name': ['Ali', 'Sara', 'Omar']
})

salaries = pd.DataFrame({
    'EmpID': [1, 2, 4],
    'Salary': [5000, 6000, 7000]
})

# Merge on common column 'EmpID'
result = pd.merge(employees, salaries, on='EmpID', how='inner')
print(result)

   EmpID  Name  Salary
0      1   Ali    5000
1      2  Sara    6000


pd.join() – Join on Index or Key Columns

In [ ]:
df1 = pd.DataFrame({'Name': ['Ali', 'Sara', 'Omar']}, index=[1, 2, 3])
df2 = pd.DataFrame({'Salary': [5000, 6000, 7000]}, index=[1, 2, 4])

# Join by index
result = df1.join(df2, how='inner')
print(result)

   Name  Salary
1   Ali    5000
2  Sara    6000


Missing Values

In [ ]:
data.isnull().sum()

id                 0
first_name         0
last_name          0
email            122
gender            43
street_num       262
street_name       37
street_suffix     37
city              79
state             80
postcode         157
dtype: int64

In [ ]:
data_products.isnull().sum()

Unnamed: 0    0
id            0
product       0
cost          0
company       5
dtype: int64

In [ ]:
data_products.isnull().sum()

Unnamed: 0    0
id            0
product       0
cost          0
company       5
dtype: int64

# Missing Values Strategies:

How to handle missing values after checking it with isnull()

In [ ]:
customer_data_copy = data.copy()
customer_data_copy.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
1   2     Cosimo   Molyneaux      cmolyneaux1@wiley.com    Male         NaN   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
4   5    Mikaela    Althorpe           malthorpe4@51.la     NaN         NaN   

  street_name street_suffix         city       state  postcode  
0   Trailsway          Road    San Diego  California   92127.0  
1         NaN           NaN      El Paso       Texas       NaN  
2    Arkansas        Circle  San Antonio       Texas   78220.0  
3        Debs         Point      Memphis   Tennessee   38150.0  
4         2nd         Drive          NaN         NaN   83732.0

Strategy 1: dropna() -> only when you have a very small proportion of the missing data. Otherwise, please avoid this strategy because that leads to the loss of information.

In [ ]:
customers_copy = customer_data_copy.dropna()
customer_data_copy.isnull().sum()

id                 0
first_name         0
last_name          0
email            122
gender            43
street_num       262
street_name       37
street_suffix     37
city              79
state             80
postcode         157
dtype: int64

In [ ]:
customers_copy.shape

(418, 11)

In [ ]:
customers_copy.head()

id first_name   last_name                      email  gender  street_num  \
0   1     Romain   Southcott  rsouthcott0@clickbank.net    Male         1.0   
2   3      Bambi     Westrip     bwestrip2@symantec.com  Female      4057.0   
3   4     Roarke  Pankettman     rpankettman3@wiley.com    Male        74.0   
5   6  Magdalena      Cullip           mcullip5@tiny.cc  Female      9190.0   
6   7   Marietta      Heball          mheball6@blog.com  Female        12.0   

  street_name street_suffix          city       state  postcode  
0   Trailsway          Road     San Diego  California   92127.0  
2    Arkansas        Circle   San Antonio       Texas   78220.0  
3        Debs         Point       Memphis   Tennessee   38150.0  
5     Packers         Drive     Baltimore    Maryland   21229.0  
6     Mallory        Center  Carol Stream    Illinois   60351.0

Strategy 2: Data Imputation

Imputation -> With the measures of central tendency [mean, median, mode]

How can we deal with the categorical data -> mode

Suppose i have 40% null data and remaining data have 50% male and 50% female and if use male to fill remaining null values as it is first occuring will that be an inaccuracy?

In [ ]:
mode_gender = data['gender'].mode()[0]
print(mode_gender)

Male


In [ ]:
data['gender'].fillna(mode_gender, inplace=True)
data.isnull().sum()

/tmp/ipython-input-3031346686.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['gender'].fillna(mode_gender, inplace=True)


id                 0
first_name         0
last_name          0
email            122
gender             0
street_num       262
street_name       37
street_suffix     37
city              79
state             80
postcode         157
dtype: int64

In [ ]:
data.shape

(1000, 11)

In [ ]:
# rename_dict = {
#    0: 'firstCustomer',
#    1: 'secondCustomer'
#}
#data.rename(index=rename_dict, inplace=True)
data.head(20)

id  first_name      last_name                        email  \
firstCustomer    1      Romain      Southcott    rsouthcott0@clickbank.net   
secondCustomer   2      Cosimo      Molyneaux        cmolyneaux1@wiley.com   
2                3       Bambi        Westrip       bwestrip2@symantec.com   
3                4      Roarke     Pankettman       rpankettman3@wiley.com   
4                5     Mikaela       Althorpe             malthorpe4@51.la   
5                6   Magdalena         Cullip             mcullip5@tiny.cc   
6                7    Marietta         Heball            mheball6@blog.com   
7                8        Tine      McSperrin  tmcsperrin7@statcounter.com   
8                9  Enrichetta     de Villier         edevillier8@ox.ac.uk   
9               10        Sari        Poulden           spoulden9@xing.com   
10              11      Natale        Martina      nmartinaa@wordpress.com   
11              12        Noni     Norrington    nnorringtonb@netvibes.com   
12              13       Mikol        MacWhan          mmacwhanc@patch.com   
13              14        Deny        Thraves            dthravesd@ibm.com   
14              15        Fara       Petegrew          fpetegrewe@ebay.com   
15              16       Jodie  De Cristoforo       jdecristoforof@mac.com   
16              17        Tobi        Dewicke      tdewickeg@hostgator.com   
17              18    Reinhold      Woolforde           rwoolfordeh@cbc.ca   
18              19      Fredek          Foley              ffoleyi@bbb.org   
19              20      Rubina       Hustings     rhustingsj@wikimedia.org   

                gender  street_num   street_name street_suffix           city  \
firstCustomer     Male         1.0     Trailsway          Road      San Diego   
secondCustomer    Male         NaN           NaN           NaN        El Paso   
2               Female      4057.0      Arkansas        Circle    San Antonio   
3                 Male        74.0          Debs         Point        Memphis   
4                 Male         NaN           2nd         Drive            NaN   
5               Female      9190.0       Packers         Drive      Baltimore   
6               Female        12.0       Mallory        Center   Carol Stream   
7                 Male       530.0          Erie         Plaza    Kansas City   
8               Female       745.0      Annamark        Street      Fairbanks   
9               Female         2.0          Pond          Hill    New Orleans   
10                Male       221.0      Sauthoff          Pass      Bradenton   
11              Female         5.0           NaN           NaN      Baltimore   
12                Male         9.0   Eagle Crest        Center          Bronx   
13              Female     71865.0   Valley Edge         Place     Washington   
14              Female         NaN     Claremont         Place    Los Angeles   
15                Male     50948.0       Rowland         Alley     Charleston   
16              Female      6817.0          Orin      Junction     San Rafael   
17                Male         NaN  Northwestern           Way       Amarillo   
18                Male         8.0      Ridgeway          Park  Silver Spring   
19              Female       249.0         Esker          Hill    San Antonio   

                               state  postcode  
firstCustomer             California   92127.0  
secondCustomer                 Texas       NaN  
2                              Texas   78220.0  
3                          Tennessee   38150.0  
4                                NaN   83732.0  
5                           Maryland   21229.0  
6                           Illinois   60351.0  
7                           Missouri   64193.0  
8                             Alaska   99709.0  
9                          Louisiana       NaN  
10                           Florida   34205.0  
11                          Maryland   21282.0  
12                          New York  

# Deal Numeric Data

How can we deal with the numeric data -> ffill(), bfill(), mean(), median()

Forward Fill - ffill() - Fills missing values using the previous non-null value (value of index 5 is copied to index of 6)

In [ ]:
data.head(40)

id  first_name      last_name                        email  \
firstCustomer    1      Romain      Southcott    rsouthcott0@clickbank.net   
secondCustomer   2      Cosimo      Molyneaux        cmolyneaux1@wiley.com   
2                3       Bambi        Westrip       bwestrip2@symantec.com   
3                4      Roarke     Pankettman       rpankettman3@wiley.com   
4                5     Mikaela       Althorpe             malthorpe4@51.la   
5                6   Magdalena         Cullip             mcullip5@tiny.cc   
6                7    Marietta         Heball            mheball6@blog.com   
7                8        Tine      McSperrin  tmcsperrin7@statcounter.com   
8                9  Enrichetta     de Villier         edevillier8@ox.ac.uk   
9               10        Sari        Poulden           spoulden9@xing.com   
10              11      Natale        Martina      nmartinaa@wordpress.com   
11              12        Noni     Norrington    nnorringtonb@netvibes.com   
12              13       Mikol        MacWhan          mmacwhanc@patch.com   
13              14        Deny        Thraves            dthravesd@ibm.com   
14              15        Fara       Petegrew          fpetegrewe@ebay.com   
15              16       Jodie  De Cristoforo       jdecristoforof@mac.com   
16              17        Tobi        Dewicke      tdewickeg@hostgator.com   
17              18    Reinhold      Woolforde           rwoolfordeh@cbc.ca   
18              19      Fredek          Foley              ffoleyi@bbb.org   
19              20      Rubina       Hustings     rhustingsj@wikimedia.org   
20              21       Karim        Woosnam          kwoosnamk@ifeng.com   
21              22     Woodrow          Fidoe  wfidoel@creativecommons.org   
22              23     Loralie        Normand            lnormandm@usa.gov   
23              24      Danita      Cleynaert      dcleynaertn@t-online.de   
24              25      Tomlin      Massinger                          NaN   
25              26    Franklin         Goodge        fgoodgep@mapquest.com   
26              27     Harlene         Rameau     hrameauq@squarespace.com   
27              28     Brander      Quinnelly    bquinnellyr@shinystat.com   
28              29      Jeanna        Bruntje      jbruntjes@tuttocitta.it   
29              30      Ardyth           Zack         azackt@wordpress.com   
30              31       Steve      Pantridge                          NaN   
31              32        Brit          Nouch            bnouchv@yahoo.com   
32              33       Caril       Barnsley    cbarnsleyw@bravesites.com   
33              34    Rosalynd        Dyerson       rdyersonx@geocities.jp   
34              35      Griffy         Sonley      gsonleyy@deviantart.com   
35              36      Ernest          Armal      earmalz@dailymail.co.uk   
36              37     Elspeth         Mendus          emendus10@ocn.ne.jp   
37              38     Clarita        Bussons   cbussons11@marketwatch.com   
38              39   Stanislas         Hacaud                          NaN   
39              40      Melvyn        Seifert   mseifert13@photobucket.com   

                gender  street_num   street_name street_suffix           city  \
firstCustomer     Male         1.0     Trailsway          Road      San Diego   
secondCustomer    Male         NaN           NaN           NaN        El Paso   
2               Female      4057.0      Arkansas        Circle    San Antonio   
3                 Male        74.0          Debs         Point        Memphis   
4                 Male         NaN           2nd         Drive            NaN   
5               Female      9190.0       Packers         Drive      Baltimore   
6               Female        12.0       Mallory        Center   Carol Stream   
7                 Male       530.0          Erie         Plaza    Kansas City   
8               Female       745.0      Annamark        Street      Fairbanks   
9      

In [ ]:
# ffill = forward fill - will update the value with one record earlier (upside value copy)

# data['street_num'] = data['street_num'].fillna(method='ffill')

data['street_num'].fillna(method='ffill', inplace=True)
data.isnull().sum()

/tmp/ipython-input-1103280876.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['street_num'].fillna(method='ffill', inplace=True)
/tmp/ipython-input-1103280876.py:3: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['street_num'].fillna(method='ffill', inplace=True)


id                 0
first_name         0
last_name          0
email            122
gender             0
street_num         0
street_name       37
street_suffix     37
city              79
state             80
postcode         157
dtype: int64

In [ ]:
# ffill = forward fill - will update the value with one record earlier (upside value copy)

data['street_suffix'].fillna(method='ffill', inplace=True)
data['street_name'].fillna(method='ffill', inplace=True)

data.isnull().sum()

/tmp/ipython-input-76712263.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['street_suffix'].fillna(method='ffill', inplace=True)
/tmp/ipython-input-76712263.py:3: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['street_suffix'].fillna(method='ffill', inplace=True)
/tmp/ipython-input-76712263.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
T

id                 0
first_name         0
last_name          0
email            122
gender             0
street_num         0
street_name        0
street_suffix      0
city              79
state             80
postcode         157
dtype: int64

In [ ]:
data.head(40)

id  first_name      last_name                        email  \
firstCustomer    1      Romain      Southcott    rsouthcott0@clickbank.net   
secondCustomer   2      Cosimo      Molyneaux        cmolyneaux1@wiley.com   
2                3       Bambi        Westrip       bwestrip2@symantec.com   
3                4      Roarke     Pankettman       rpankettman3@wiley.com   
4                5     Mikaela       Althorpe             malthorpe4@51.la   
5                6   Magdalena         Cullip             mcullip5@tiny.cc   
6                7    Marietta         Heball            mheball6@blog.com   
7                8        Tine      McSperrin  tmcsperrin7@statcounter.com   
8                9  Enrichetta     de Villier         edevillier8@ox.ac.uk   
9               10        Sari        Poulden           spoulden9@xing.com   
10              11      Natale        Martina      nmartinaa@wordpress.com   
11              12        Noni     Norrington    nnorringtonb@netvibes.com   
12              13       Mikol        MacWhan          mmacwhanc@patch.com   
13              14        Deny        Thraves            dthravesd@ibm.com   
14              15        Fara       Petegrew          fpetegrewe@ebay.com   
15              16       Jodie  De Cristoforo       jdecristoforof@mac.com   
16              17        Tobi        Dewicke      tdewickeg@hostgator.com   
17              18    Reinhold      Woolforde           rwoolfordeh@cbc.ca   
18              19      Fredek          Foley              ffoleyi@bbb.org   
19              20      Rubina       Hustings     rhustingsj@wikimedia.org   
20              21       Karim        Woosnam          kwoosnamk@ifeng.com   
21              22     Woodrow          Fidoe  wfidoel@creativecommons.org   
22              23     Loralie        Normand            lnormandm@usa.gov   
23              24      Danita      Cleynaert      dcleynaertn@t-online.de   
24              25      Tomlin      Massinger                          NaN   
25              26    Franklin         Goodge        fgoodgep@mapquest.com   
26              27     Harlene         Rameau     hrameauq@squarespace.com   
27              28     Brander      Quinnelly    bquinnellyr@shinystat.com   
28              29      Jeanna        Bruntje      jbruntjes@tuttocitta.it   
29              30      Ardyth           Zack         azackt@wordpress.com   
30              31       Steve      Pantridge                          NaN   
31              32        Brit          Nouch            bnouchv@yahoo.com   
32              33       Caril       Barnsley    cbarnsleyw@bravesites.com   
33              34    Rosalynd        Dyerson       rdyersonx@geocities.jp   
34              35      Griffy         Sonley      gsonleyy@deviantart.com   
35              36      Ernest          Armal      earmalz@dailymail.co.uk   
36              37     Elspeth         Mendus          emendus10@ocn.ne.jp   
37              38     Clarita        Bussons   cbussons11@marketwatch.com   
38              39   Stanislas         Hacaud                          NaN   
39              40      Melvyn        Seifert   mseifert13@photobucket.com   

                gender  street_num   street_name street_suffix           city  \
firstCustomer     Male         1.0     Trailsway          Road      San Diego   
secondCustomer    Male         1.0     Trailsway          Road        El Paso   
2               Female      4057.0      Arkansas        Circle    San Antonio   
3                 Male        74.0          Debs         Point        Memphis   
4                 Male        74.0           2nd         Drive            NaN   
5               Female      9190.0       Packers         Drive      Baltimore   
6               Female        12.0       Mallory        Center   Carol Stream   
7                 Male       530.0          Erie         Plaza    Kansas City   
8               Female       745.0      Annamark        Street      Fairbanks   
9      

Backward Fill - bfill() - Fills missing values using the next non-null value (value of index 6 is copied to index of 5)

In [ ]:
data['city'].fillna(method='bfill', inplace=True)
data['state'].fillna(method='bfill', inplace=True)
data['postcode'].fillna(method='bfill', inplace=True)

data.head(40)

/tmp/ipython-input-1775249522.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['city'].fillna(method='bfill', inplace=True)
/tmp/ipython-input-1775249522.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['city'].fillna(method='bfill', inplace=True)
/tmp/ipython-input-1775249522.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior 

id  first_name      last_name                        email  \
firstCustomer    1      Romain      Southcott    rsouthcott0@clickbank.net   
secondCustomer   2      Cosimo      Molyneaux        cmolyneaux1@wiley.com   
2                3       Bambi        Westrip       bwestrip2@symantec.com   
3                4      Roarke     Pankettman       rpankettman3@wiley.com   
4                5     Mikaela       Althorpe             malthorpe4@51.la   
5                6   Magdalena         Cullip             mcullip5@tiny.cc   
6                7    Marietta         Heball            mheball6@blog.com   
7                8        Tine      McSperrin  tmcsperrin7@statcounter.com   
8                9  Enrichetta     de Villier         edevillier8@ox.ac.uk   
9               10        Sari        Poulden           spoulden9@xing.com   
10              11      Natale        Martina      nmartinaa@wordpress.com   
11              12        Noni     Norrington    nnorringtonb@netvibes.com   
12              13       Mikol        MacWhan          mmacwhanc@patch.com   
13              14        Deny        Thraves            dthravesd@ibm.com   
14              15        Fara       Petegrew          fpetegrewe@ebay.com   
15              16       Jodie  De Cristoforo       jdecristoforof@mac.com   
16              17        Tobi        Dewicke      tdewickeg@hostgator.com   
17              18    Reinhold      Woolforde           rwoolfordeh@cbc.ca   
18              19      Fredek          Foley              ffoleyi@bbb.org   
19              20      Rubina       Hustings     rhustingsj@wikimedia.org   
20              21       Karim        Woosnam          kwoosnamk@ifeng.com   
21              22     Woodrow          Fidoe  wfidoel@creativecommons.org   
22              23     Loralie        Normand            lnormandm@usa.gov   
23              24      Danita      Cleynaert      dcleynaertn@t-online.de   
24              25      Tomlin      Massinger                          NaN   
25              26    Franklin         Goodge        fgoodgep@mapquest.com   
26              27     Harlene         Rameau     hrameauq@squarespace.com   
27              28     Brander      Quinnelly    bquinnellyr@shinystat.com   
28              29      Jeanna        Bruntje      jbruntjes@tuttocitta.it   
29              30      Ardyth           Zack         azackt@wordpress.com   
30              31       Steve      Pantridge                          NaN   
31              32        Brit          Nouch            bnouchv@yahoo.com   
32              33       Caril       Barnsley    cbarnsleyw@bravesites.com   
33              34    Rosalynd        Dyerson       rdyersonx@geocities.jp   
34              35      Griffy         Sonley      gsonleyy@deviantart.com   
35              36      Ernest          Armal      earmalz@dailymail.co.uk   
36              37     Elspeth         Mendus          emendus10@ocn.ne.jp   
37              38     Clarita        Bussons   cbussons11@marketwatch.com   
38              39   Stanislas         Hacaud                          NaN   
39              40      Melvyn        Seifert   mseifert13@photobucket.com   

                gender  street_num   street_name street_suffix           city  \
firstCustomer     Male         1.0     Trailsway          Road      San Diego   
secondCustomer    Male         1.0     Trailsway          Road        El Paso   
2               Female      4057.0      Arkansas        Circle    San Antonio   
3                 Male        74.0          Debs         Point        Memphis   
4                 Male        74.0           2nd         Drive      Baltimore   
5               Female      9190.0       Packers         Drive      Baltimore   
6               Female        12.0       Mallory        Center   Carol Stream   
7                 Male       530.0          Erie         Plaza    Kansas City   
8               Female       745.0      Annamark        Street      Fairbanks   
9      

In [ ]:
data.isnull().sum()

id                 0
first_name         0
last_name          0
email            122
gender             0
street_num         0
street_name        0
street_suffix      0
city               0
state              0
postcode           0
dtype: int64

We have 122 records of email whose value is not available, and we cannot alter this data as it is quite sensitive data so in that case we can fill it with some value which indicates us that value of particular customer is not available.

In [ ]:
data['email'].fillna('Unknown', inplace=True)
data.head(40)

/tmp/ipython-input-4250169982.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['email'].fillna('Unknown', inplace=True)


id  first_name      last_name                        email  \
firstCustomer    1      Romain      Southcott    rsouthcott0@clickbank.net   
secondCustomer   2      Cosimo      Molyneaux        cmolyneaux1@wiley.com   
2                3       Bambi        Westrip       bwestrip2@symantec.com   
3                4      Roarke     Pankettman       rpankettman3@wiley.com   
4                5     Mikaela       Althorpe             malthorpe4@51.la   
5                6   Magdalena         Cullip             mcullip5@tiny.cc   
6                7    Marietta         Heball            mheball6@blog.com   
7                8        Tine      McSperrin  tmcsperrin7@statcounter.com   
8                9  Enrichetta     de Villier         edevillier8@ox.ac.uk   
9               10        Sari        Poulden           spoulden9@xing.com   
10              11      Natale        Martina      nmartinaa@wordpress.com   
11              12        Noni     Norrington    nnorringtonb@netvibes.com   
12              13       Mikol        MacWhan          mmacwhanc@patch.com   
13              14        Deny        Thraves            dthravesd@ibm.com   
14              15        Fara       Petegrew          fpetegrewe@ebay.com   
15              16       Jodie  De Cristoforo       jdecristoforof@mac.com   
16              17        Tobi        Dewicke      tdewickeg@hostgator.com   
17              18    Reinhold      Woolforde           rwoolfordeh@cbc.ca   
18              19      Fredek          Foley              ffoleyi@bbb.org   
19              20      Rubina       Hustings     rhustingsj@wikimedia.org   
20              21       Karim        Woosnam          kwoosnamk@ifeng.com   
21              22     Woodrow          Fidoe  wfidoel@creativecommons.org   
22              23     Loralie        Normand            lnormandm@usa.gov   
23              24      Danita      Cleynaert      dcleynaertn@t-online.de   
24              25      Tomlin      Massinger                      Unknown   
25              26    Franklin         Goodge        fgoodgep@mapquest.com   
26              27     Harlene         Rameau     hrameauq@squarespace.com   
27              28     Brander      Quinnelly    bquinnellyr@shinystat.com   
28              29      Jeanna        Bruntje      jbruntjes@tuttocitta.it   
29              30      Ardyth           Zack         azackt@wordpress.com   
30              31       Steve      Pantridge                      Unknown   
31              32        Brit          Nouch            bnouchv@yahoo.com   
32              33       Caril       Barnsley    cbarnsleyw@bravesites.com   
33              34    Rosalynd        Dyerson       rdyersonx@geocities.jp   
34              35      Griffy         Sonley      gsonleyy@deviantart.com   
35              36      Ernest          Armal      earmalz@dailymail.co.uk   
36              37     Elspeth         Mendus          emendus10@ocn.ne.jp   
37              38     Clarita        Bussons   cbussons11@marketwatch.com   
38              39   Stanislas         Hacaud                      Unknown   
39              40      Melvyn        Seifert   mseifert13@photobucket.com   

                gender  street_num   street_name street_suffix           city  \
firstCustomer     Male         1.0     Trailsway          Road      San Diego   
secondCustomer    Male         1.0     Trailsway          Road        El Paso   
2               Female      4057.0      Arkansas        Circle    San Antonio   
3                 Male        74.0          Debs         Point        Memphis   
4                 Male        74.0           2nd         Drive      Baltimore   
5               Female      9190.0       Packers         Drive      Baltimore   
6               Female        12.0       Mallory        Center   Carol Stream   
7                 Male       530.0          Erie         Plaza    Kansas City   
8               Female       745.0      Annamark        Street      Fairbanks   
9      

In [ ]:
data.isnull().sum()

id               0
first_name       0
last_name        0
email            0
gender           0
street_num       0
street_name      0
street_suffix    0
city             0
state            0
postcode         0
dtype: int64

Task: Impute the missing values of the company with the 'Unknown'

In [ ]:
data_products.isnull().sum()

Unnamed: 0    0
id            0
product       0
cost          0
company       5
dtype: int64

In [ ]:
data_products['company'].fillna('Unknown', inplace=True)
data_products.isnull().sum()

/tmp/ipython-input-3886835828.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data_products['company'].fillna('Unknown', inplace=True)


Unnamed: 0    0
id            0
product       0
cost          0
company       0
dtype: int64

In [ ]:
data.to_excel('customers.xlsx')
data_purchases.to_excel('purchases.xlsx')
data_products.to_excel('products.xlsx')

# Data Analysis: Aggregation

Group by 'customer_num' and compute the sum of 'amount' and 'paid' column

In [ ]:
data_purchases.head(10)

Unnamed: 0  id           purch_date  customer_num  product_num  amount  \
0           0   1  2019-01-03 00:00:00           823           27      12   
1           1   2  2019-01-03 00:00:00           606           28      14   
2           2   3  2019-01-03 00:00:00           955            9      17   
3           3   4  2019-01-03 00:00:00           577           19       3   
4           4   5  2019-01-03 00:00:00           429            8      18   
5           5   6  2019-01-03 00:00:00           275           36      18   
6           6   7  2019-01-03 00:00:00           456           21      10   
7           7   8  2019-01-03 00:00:00           526           58      19   
8           8   9  2019-01-03 00:00:00           326           44       2   
9           9  10  2019-01-03 00:00:00            13           21       5   

        paid  
0    $568.92  
1    $395.36  
2    $510.17  
3     $68.49  
4    $759.42  
5    $176.76  
6    $904.80  
7  $1,387.95  
8     $61.34  
9    $316.70

In [ ]:
aggregated_purchase = data_purchases.groupby('customer_num').agg({'amount':'sum', 'paid': 'sum'})
print(aggregated_purchase)

              amount                                               paid
customer_num                                                           
1                 24                         $62.80$22.67$326.75$367.12
2                 67     $1,215.62$260.32$174.10$107.35$1,809.60$411.10
3                 88  $1,262.76$342.45$592.30$1,392.80$168.84$304.02...
4                 38                       $743.88$743.88$212.80$575.82
5                 46               $749.40$1,252.90$56.48$651.36$221.10
...              ...                                                ...
996               31           $61.99$42.99$84.38$1,690.81$21.47$437.55
997               55       $187.80$406.90$793.12$143.80$1,339.52$159.20
998               69        $1,179.20$168.90$548.47$84.78$724.80$640.78
999               66      $113.89$642.43$557.70$1,075.88$74.28$1,176.30
1000              75  $289.92$203.45$617.67$65.35$600.20$542.34$1,12...

[997 rows x 2 columns]


Task: Find the total amount and total paid as per the purchase_date.

In [ ]:
aggregated_purchase_date = data_purchases.groupby('purch_date').agg({'amount':'sum', 'paid':'sum'})
aggregated_purchase_date

amount                                               paid
purch_date                                                                    
2019-01-03 00:00:00     480  $568.92$395.36$510.17$68.49$759.42$176.76$904....
2019-01-04 00:00:00     872  $125.04$125.24$180.96$457.70$1,006.91$459.18$4...
2019-01-05 00:00:00     715  $302.32$1,678.86$808.64$1,692.18$375.12$180.06...
2019-01-06 00:00:00     576  $208.17$480.42$755.80$8.95$853.92$480.16$1,046...
2019-02-03 00:00:00     568  $150.29$135.21$107.26$136.98$165.96$1,215.62$7...
...                     ...                                                ...
06/16/2019              410  $229.20$307.56$220.32$482.00$493.20$425.60$162...
06/17/2019              376  $343.86$70.70$870.50$61.34$261.40$1,090.56$377...
06/18/2019              406  $211.95$379.71$660.06$818.35$674.10$332.70$235...
06/19/2019              565  $180.78$1,013.94$68.40$1,662.69$929.85$168.76$...
06/20/2019              458  $297.12$109.32$518.64$443.60$111.15$96.81$122....

[108 rows x 2 columns]

Task: Please explore the method drop_duplicates

drop_duplicates() removes duplicate rows (or specific column combinations) from a DataFrame.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Name': ['Ali', 'Sara', 'Ali', 'Zara', 'Sara'],
    'Age': [25, 30, 25, 22, 30]
})

print("Original:")
print(df)

cleaned = df.drop_duplicates()
print("\nAfter drop_duplicates:")
print(cleaned)

Original:
   Name  Age
0   Ali   25
1  Sara   30
2   Ali   25
3  Zara   22
4  Sara   30

After drop_duplicates:
   Name  Age
0   Ali   25
1  Sara   30
3  Zara   22
